### DISCUSSION TITLE
Basic use case, onsuccess redirect in atomsWithMutation (nextjs)

### DISCUSSION THREAD

**Discussion by lveillard** [pmndrs/jotai#2037] [Q&A] [answered]
Hello! So i'm migrating from the hooks world of a nextjs app to the atoms world, and besides some hydration issues i was able to do almost everything, Two things i'm struggling a bit with:

1) My former useMutation worked like this:
```
export const useMutateThings = (onSuccess?: () => void) => {
  // const queryClient = useQueryClient();

  const result = useMutation({
    mutationFn: (data: {
      payload: any;
    }) => {
      const res = mutate(data);
      return res;
    },
    onSuccess: () => {
      if (onSuccess) {
        onSuccess();
      }
    },
  });

  return result;
};
```

This enabled me to in each instance using this hook, decide the onSuccess.  Including a redirection. 

I'm completely unable to do it for the atomWithMutation. Can't use hooks inside the atom, and can't send onSuccess function for the onSuccess calback,

2) My second issue is with the queries (tanstack queries atom). Sometimes i can set an atom and use it, but other times is not like a state but like a param. For instance having N components each one with its state and firing queries for them, i could store all the ids in one atom but not the perfect solution for my use case. 

In the same topic, I can't set the right key for each atom as i can't send it as a param.

I think both issues open the same question: Is atomFamily the only door to send params to atoms? I am forced to do atomFamilies of my queries and mutations if i can't store the params as atoms?

**Comment by lveillard** [comment]
Some weird workarounds i'm trying for issue 1:

Workaround A:
```
  const [mutation, mutateFunction] = useAtom(myMutationAtom);
  const { data , isLoading} = mutation;
  
  useEffect(() => {
  if (data) {
    router.push("/new/path");
  }
}, [data]);
```
Problem is that data does not get removed so if you come back to the same component, the redirection will happen again.

Workaround B:
This one is uglier and not sure if it works
```
//myatom.ts
export const [asyncData, mutateUnitAtom] = atomsWithMutation(

//MyComp.tsx

const MyComp = () => 
... stuff
<Suspense>
    <Redirection>
 </Suspense>
...

const Refirection = > {
  useEffect(() => {
    router.push("/new/path");
}, []);
}
```

As Redirection is consuming an async atom, is not being rendered. Once the data arrives, it gets rendered and the router.push happens 😅


For the issue 2 I still don't have any workaround

**Comment by dai-shi** [comment, marked as answer]
First of all, I don't think atomsWithMutation is well designed for DX, and it would be nice if someone can re-design jotai-tanstack-query.
Second of all, atoms and props are not friends. If you need props in atoms, you would need to create an atom in render.

```ts
export const useMutateThings = (onSuccess?: () => void) => {
  const onSuccessAtom = useRef(atom({ onSuccess })).current;
  const setOnSuccess = useSetAtom(onSuccessAtom);
  useEffect(() => setOnSuccess({ onSuccess });

  const [, statusAtom] = useRef(atomsWithMutation((get) => ({
    // ...
    onSuccess: get(onSuccessAtom).onSuccess,
  }))).current;

  const [status, dispatch] = useAtom(statusAtom);
  useEffect(() => dispatch(...), [...]);
  return status.data;
};
```

- I don't know if the code is correct.
- atomsWithMutation seems very different from useMutation.
- I would use `use-constant` instead of `useRef`.

Hope it helps a bit.

**Reply by lveillard** [reply]
It helps, thanks! 
I think it makes sense to have the queries as atoms to spread dependencies, but maybe I will keep the mutations as hooks.

**Comment by lveillard** [comment]
Probably not the cleanest option, but i'm exploring having queries as atoms and mutations as hooks.

In the mutation hooks, I gather all the atoms I require and I pass them to the asyncMutation. it looks ugly and naive, probably there is a better way of doing but it seems to work without infinite rerenders and is easy to read. 

Might help somebody, the pattern looks like this:

```
///useMutateRecord.ts
export const useMutateRecord = (recordId: string, onSuccess?: () => void) => {
  const currentContextId= useAtomValue(currentContextIdAtom);
  const dataDeltas= useAtomValue(dataDeltasAtom(recordId));

  const result = useMutation({
    mutationKey: ['record', recordId],
    mutationFn: async () => {
        /// mutation logic      
        ///...
        return { unitId, spaceId, arcDeltas };
    },

    onSuccess: (result) => {
      if (onSuccess) {
        onSuccess(result);
      }
    },
  });
  ```
  
  And then in my component logic:
  ```
  /// MyComponent.tsx
     ///...
    const { mutateAsync: mutateCurrentRecord } = useMutateRecordId(
    recordId,
    (result) => {router.push(`/record/${result.id}`)}
    });
    
    return ( 
    
    <>
      ...
     <Button  onClick={mutateCurrentRecord} > Create Record </Button>
    </>
    )
```

**Comment by lveillard** [comment]
And something i feel it is a bit more clean is just to setup a provider with a store, and store.get(yourAtoms) directly in the mutationFn function, so those atoms don't become dependencies of the component, and are just read when clicking the submit button. 

Happy to hear other alternatives tho!

**Comment by lveillard** [comment]
So summarizing the pain points:
- No params ( so no way to assign a query to the query, or query ids that can't / should not be stored in toms)
- No hooks inside atoms, so no way to run onSuccess directly on the query / mutate atoms

My use case:
- I run a query on component mount
- Then i push that data into an atom cache onSuccess
- When running a query from another source (for instance invalidating that query) the cache must be automatically reupdated

Two other ugly alternatives:

### Option C
Hook that creates memoized atoms and receives params. 
Disadvantage => Can't easily set atoms onSuccess
```

export const useLoadRecordAtom = (recordId: string, payload, onSuccess) => {
  return useMemo(
    () =>
      atomsWithQuery((get) => {
        const spaceId = get(spaceIdAtom);

        return {
          queryKey: [recordId],
          queryFn: async () => query(recordId, payload),
          onSuccess,
        };
      }),
    [recordId, onSuccess]
  );
};
```

### Option D
- Create a write atom and use the params of the write atom in the query atom inside (it "works", but looks risky, not gonna lie)
```
export const cheatQuery= atom(
  null,
  async (get, set, { recordId, payload, onSuccess }) => {
    const [loadRecordAtomTwo] = atomsWithQuery((g) => 

      return {
        queryKey: [recordId],
        queryFn: async () => query(recordId,payload),
        onSuccess: (data) => {
            set(setRecorCacheAtom, data; // here we are dangerously setting atoms onSuccess
            onSuccess(data); // This will run the router.path
        },
      };
    });
    return await get(loadRecordAtomTwo);
);
```

**Reply by dai-shi** [reply]
For option D, I don't think you need `atomsWithQuery`. You can do `const client = get(clientAtom)` and use methods on `client`.

**Reply by shongs27** [reply]
@dai-shi 
Could you give me a simple example? How can I change the value of another atom in onSuccess without atom in Option D

```

const testAtom = atom(null);

export const [, changeCoach] = atomsWithMutation<
  AxiosResponse,
  AxiosError,
  ICoachWindow.CoachRequest
>((get) => ({
  mutationKey: ['coachPatch', get(companyIdsAtom)],
  mutationFn: async ({ uid, req }) => {
    const companyIds = encodeURIComponent(get(companyIdsAtom).join(','));
    const headers = get(headerInfo);

    return await fetcher({
      ...headers,
      method: 'patch',
      url: `/company/${companyIds}/user/${uid}/coach`,
      data: req,
    });
  },
  onSuccess: () => {
    get(queryClientAtom).invalidateQueries([
      `/company/${get(companyIdsAtom).join(',')}/user/target`,
      get(targetInquireConditionsAtoms),
      get(headerInfo),
    ]);
   set(testAtom, true) // => ???? how to set
  },
}));
```
this is what i want to being

**Reply by dai-shi** [reply]
It's been changed since then.
Let's see @kalijonn if they have suggestions.

**Reply by kalijonn** [reply]
for anyone else following, this moved here: https://github.com/pmndrs/jotai/discussions/2338
